src/shared/fdset.c

```diff
@@ -265,13 +280,42 @@ int fdset_to_array(FDSet *fds, int **ret) {
         if (!a)
                 return -ENOMEM;
 
-        SET_FOREACH(e, MAKE_SET(fds))
-                a[j++] = PTR_TO_FD(e);
+        for (size_t i = 0; i < m; ++i)
+                a[i] = -EBADF;
+
+        if (indexed) {
+                void *k, *e;
+
+                HASHMAP_FOREACH_KEY(e, k, MAKE_HASHMAP(fds)) {
+                        int fd = PTR_TO_FD(k), index = -(1 + PTR_TO_INT(e));
```

Comment:
can you please add a macro that does the index-to-ptr conversion? this is just too magic